mkosi/__init__.py

```diff
@@ -2755,23 +2814,35 @@ def jj(line: str) -> str:
 
 
 def install_boot_loader_clear(args: CommandLineArguments, root: str, loopdev: str) -> None:
-    nspawn_params = [
-        # clr-boot-manager uses blkid in the device backing "/" to
-        # figure out uuid and related parameters.
-        f"--bind-ro={loopdev}",
-        f"--bind-ro=/dev/block",
-        f"--bind-ro=/dev/disk",
-        f"--property=DeviceAllow={loopdev}",
-    ]
-
-    for partno in (args.esp_partno, args.bios_partno, args.root_partno, args.xbootldr_partno):
-        if partno is not None:
-            p = partition(loopdev, partno)
-            nspawn_params += [f"--bind-ro={p}", f"--property=DeviceAllow={p}"]
+    # clr-boot-manager uses blkid in the device backing "/" to
+    # figure out uuid and related parameters.
+    nspawn_params = nspawn_params_for_blockdev_access(args, loopdev)
 
     cmdline = ["/usr/bin/clr-boot-manager", "update", "-i"]
     run_workspace_command(args, root, cmdline, nspawn_params=nspawn_params)
 
+def install_boot_loader_centos_old_efi(args: CommandLineArguments, root: str, loopdev: str) -> None:
+    nspawn_params = nspawn_params_for_blockdev_access(args, loopdev)
+
+    # prepare EFI directory on ESP
+    os.makedirs(os.path.join(root, 'efi/EFI/centos'), exist_ok=True)
+
+    # patch existing or create minimal GRUB_CMDLINE config
+    write_grub_config(args, root)
+
+    # generate grub2 efi boot config
+    cmdline = ['/sbin/grub2-mkconfig', "-o", "/efi/EFI/centos/grub.cfg"]
```

Comment:
How does grub decide between BIOS and UEFI? Does it just look at the directory it's being installed in?

Reply:
As stated in the comment, it actually does a check for a directory existence of `/sys/firmware/efi`, which fails at least on my build host due to the underlying OS not being booted via UEFI. This in turn causes `grub2-mkconfig` to fall back into `16bit` mode, generating the `linux16 ...` and `initrd16 ...` bootconfig lines.
That is the reason why we need to patch it afterwards.
I've searched a bit through the various manpages and `--help` outputs for the `grub2` utilities but haven't found anything that would allow explicitly forcing an install method while ignoring/overriding the internal checks. IIRC they are in `/etc/grub.d/10-linux`.